src/orange/surf/detail/InvoluteSolver.hh

```diff
@@ -0,0 +1,306 @@
+ */
+CELER_FUNCTION auto
+InvoluteSolver::operator()(Real3 const& pos,
+                           Real3 const& dir,
+                           SurfaceState on_surface) const -> Intersections
+{
+    using constants::pi;
+    // Flatten pos and dir in xyz and uv respectively
+    real_type x = pos[0];
+    real_type const y = pos[1];
+
+    real_type u = dir[0];
+    real_type v = dir[1];
+
+    // Mirror systemm for counterclockwise involutes
+    if (sign_)
+    {
+        x = -x;
+        u = -u;
+    }
+
+    /*
+     * Define tolerances.
+     * tol_point gives the tolerance level for a point,
+     * account for the floating point error when performing square roots.
+     * tol_conv gives the tolerance for the Regular Falsi iteration.
+     */
+    real_type tol_conv = r_b_ * tol();
+    real_type tol_point = r_b_ * tol() * 100;
+
+    // Results initalization and root counter
+    Intersections result;
+    int j = 0;
+    // Initial result vector.
+    result = {no_intersection(), no_intersection(), no_intersection()};
+
+    // Return result if particle is travelling along z-axis.
+    if (u == 0 && v == 0)
+    {
+        return result;
+    }
+
+    // Conversion constant for 2-D distance to 3-D distance
+
+    real_type convert = 1 / std::sqrt(ipow<2>(v) + ipow<2>(u));
+    u *= convert;
+    v *= convert;
+
+    // Line angle parameter
+
+    real_type beta = line_angle_param(u, v);
+
+    // Setting first interval bounds, needs to be done to ensure roots are
+    // found
+    real_type t_lower = 0;
+    real_type t_upper = beta - a_;
+
+    // Round t_upper to the first positive multiple of pi
+    t_upper += max<real_type>(0, -std::floor(t_upper / pi)) * pi;
+
+    // Slow down factor to increment bounds when a root cannot be found
+    int i = 1;
+
+    // Lambda used for calculating the roots using Regula Falsi Iteration
+    auto calc_t_intersect = [&](real_type t) {
+        real_type alpha = u * std::sin(t + a_) - v * std::cos(t + a_);
+        real_type beta = t * (u * std::cos(t + a_) + v * std::sin(t + a_));
+        return r_b_ * (alpha - beta) + x * v - y * u;
+    };
+    IllinoisRootFinder find_root_between{calc_t_intersect, tol_conv};
+
+    // Iterate on roots
+    while (t_lower < tmax_)
+    {
+        // Find value in root function
+        real_type ft_lower = calc_t_intersect(t_lower);
+        real_type ft_upper = calc_t_intersect(t_upper);
+
+        // Only iterate when roots have different signs
+        if (signum<real_type>(ft_lower) != signum<real_type>(ft_upper))
+        {
+            // Regular Falsi Iteration: Sometimes will slowly converge
+            real_type t_gamma = find_root_between(t_lower, t_upper);
+
+            // Convert root to distance and store if positive and in interval
+            real_type dist = calc_dist(x, y, u, v, t_gamma);
+            if (dist > tol_point
+                || (!(on_surface == SurfaceState::on) && (dist > 0)))
```

Comment:
Better to write the more common case first:
```suggestion
            if ((dist > 0 && on_surface != SurfaceState::on)
                || dist > tol_point)
```
you could also combine this expression with
```suggestion
            if (dist > (on_surface == SurfaceState::on ? tol_point : 0))
```
which actually, since the `on_surface` and `tol_point` are loop invariant, means you should just set
```c++
tol_point = (on_surface == SurfaceState::on ? r_b_ * tol() * 100 : 0);
```
 outside of the loop, and simplify this comparison to just
```suggestion
            if (dist > tol_point)
```